Merge py3-upgrade into master (#24)
public-people/scrape-news: README.md

See the _In Progress_ column at [https://trello.com/b/9TVRB4gb/public-people](https://trello.com/b/9TVRB4gb/public-people) to see which publications are currently being tackled to avoid duplication.

Next, go get started at [Development](#development)

## Copyright of news content

We do not make news content available for public consumption. We simply store and index the news content and the original URL and publication date to provide search functionality similar to search engines. This project intends to provide better access to news on the publisher's website. It should be used to send readers to relevant news websites rather than to replace them.

## Requirements

- Python 3.5+
- Python virtual environment with packages installed from [requirements.txt](/requirements.txt). This main library used [Scrapy](https://scrapy.org/).
- Setup [scrapyd](https://scrapyd.readthedocs.io/en/stable/) (optional).
- Aleph account and credentials for uploading results (optional).

## Development

### Set up your development environment

#### Clone

Fork this repository on GitHub and clone your fork:

```bash

> $ git clone https://github.com/your-name/scrape-news.git

Reviewer: The problem with these $s is that you can't just copy and paste instructions. It's already a code block: it's clear it's code.

```suggestion
git clone https://github.com/your-name/scrape-news.git
```

```
or

```bash

> $ git clone git@github.com:your-name/scrape-news.git

Reviewer: ```suggestion
git clone git@github.com:your-name/scrape-news.git
```

```

(Make sure to replace ```your-name```.)

Navigate to the repo.

```bash
$ cd <PATH_TO_REPO>
```

### Make commands

> This project comes with a [Makefile](/Makefile), to run with `make` command. Run the following without any arguments to get a list of available commands.

Reviewer: ```suggestion
This project comes with a [Makefile](/Makefile), to run with `make` command.
```

```bash
$ make

> ```

Reviewer: Please remove.

    ```

### Run

Note: Always activate the project's environment before using it.

```bash
$ source venv/bin/activate
```

#### List available spiders in the project

```bash
$ make list
businesslivecrawl
businesslivesitemap
dailymaverick
dailyvoice
dfa
dispatchlive
enca
ewn
groundup
iol
mg
news24
rekordnorth
sabc
sowetanlive
thenewage
timeslive
```

We will use these crawler names in the next step. Note that these names are generated from _classes_ in the [spiders](/scrapenews/spiders/) directory and not the _filenames_.

#### Run scraper

Run a scraper using the command below to check that your environment is working properly. This can be done from the project root because of how the `scrapy` library works.

If you need to get a template crawl command quickly and then fill in with a crawler, run the following and then copy and paste the result to a new line.

```bash
$ make quickstart
scrapy crawl -s ITEM_PIPELINES="{}" -a since_lastmod=2018-01-01 <CRAWLER>
```

For example, run the command using the _iol_ spider.

```bash
$ scrapy crawl -s ITEM_PIPELINES="{}" -a since_lastmod=2018-04-30 iol
```

The arguments for the above command are required. Here is how to use them:

> - The setting `ITEM_PIPELINES` disables the pipeline we have configured which you don't need for just developing a spider.

Reviewer: ```suggestion
- The setting `ITEM_PIPELINES="{}"` disables the default pipeline we which you don't need for just developing a spider.
```

- The argument `since_lastmod` is the earliest sitemap file and page the scraper will include.

> - The last argument `crawl` is the name of the scraper (e.g. `iol`). See output from the previous section.

Reviewer: ```suggestion
- The last argument to the `crawl` command is the name of the scraper (e.g. `iol`). See output from the previous section.
```

For quick testing on the _iol_ spider, a shortcut for the above command has been added to the [Makefile](/Makefile).

```bash
$ make test-iol

> ```

Reviewer: Let's remove this stuff while removing it from the makefile

#### Output

> If the crawl command is working correctly, it will output a lot of information.

Reviewer: ```suggestion
If the crawl command above is working correctly, it will output a lot of information:
```

 'url': 'https://www.iol.co.za/personal-finance/stanlib-may-further-reduce-fund-offering-14717297'}
```

### Code checks

> To maintain Python code quality, **linting** and **unit tests** should be run locally and on Github. Linting give you style warnings. The unit tests are in the [unit](/tests/unit) directory and run tests against the project code.

Reviewer: ```suggestion
To maintain Python code quality, **linting** and **unit tests** should be run locally and in Continuous Integration. Linting give you style warnings. The unit tests are in the [unit](/tests/unit) directory and run tests against the project code.
```

Ideally spiders should be driven from the outlet's sitemap. Ideally you'll find the sitemap from /robots.txt. If you don't find it there, try /sitemap.xml or /sitemap.txt.

> If the ```newssite``` uses a _useful_ sitemap index (see for example [https://www.timeslive.co.za/sitemap/](https://www.timeslive.co.za/sitemap/)), use a sitemap spider.

Reviewer: It's not a technical term, just a typo :grin: 

```suggestion
If the news site uses a _useful_ sitemap index (see for example [https://www.timeslive.co.za/sitemap/](https://www.timeslive.co.za/sitemap/)), use a sitemap spider.
```

### Test your responses

To test individual xpath or css responses you can use the scrapy shell:

```bash
scrapy shell "https://www.newssite.co.za/article-url"
```

> If you go to the same url in your browser and right-click on, say, the title of the article, and select 'Inspect Element (Q)', you'll see something like this highlighted.

Reviewer: ```suggestion
If you go to the same url in your browser and right-click on, say, the title of the article, and select 'Inspect Element (Q)', you'll see something like this highlighted:
```
